src/placeos-rest-api/controllers/root.cr

```diff
@@ -77,6 +84,51 @@ module PlaceOS::Api
       )
     end
 
+    SERVICES = %w(frontend rubber core triggers dispatch)
+
+    def self.construct_versions : Array(PlaceOS::Model::Version)
+      version_channel = Channel(PlaceOS::Model::Version?).new
+      {% for service in SERVICES %}
+        spawn do
+          %version = begin
+                      {{service.id}}_version
+                    rescue
+                      Log.warn { {service: {{ service }}, message: "failed to request version" }}
+                      nil
+                    end
+          version_channel.send(%version)
+          end
+        
+      {% end %}
+      Array(PlaceOS::Model::Version?).new(SERVICES.size) do
+        version_channel.receive
+      end.compact
+    end
+
+    private def self.frontend_version : (PlaceOS::Model::Version | Nil)
+      Frontends::Client.client(&.version)
+    end
+
+    private def self.rubber_version : (PlaceOS::Model::Version | Nil)
+      RubberSoul::Client.client(&.version)
+    end
+
+    private def self.core_version : (PlaceOS::Model::Version | Nil)
+      Core::Client.client(PLACE_URI, &.version)
+    end
+
+    private def self.triggers_version : (PlaceOS::Model::Version | Nil)
+      trigger_uri = TRIGGERS_URI.dup
+      trigger_uri.path = "/api/triggers/v2/version"
+      response = HTTP::Client.get trigger_uri
+      PlaceOS::Model::Version.from_json(response.body)
+    end
+
+    private def self.dispatch_version : (PlaceOS::Model::Version | Nil)
+      response = HTTP::Client.get "#{PLACE_URI}/api/server/version"
+      PlaceOS::Model::Version.from_json(response.body)
```

Comment:
Since there's no error handling, this actually won't return `nil`, it will just raise if there's a problem